actors/builtin/multisig/multisig_actor.go

```diff
@@ -355,44 +364,57 @@ func (a Actor) ChangeNumApprovalsThreshold(rt vmr.Runtime, params *ChangeNumAppr
 	return nil
 }
 
-func (a Actor) approveTransaction(rt vmr.Runtime, txnID TxnID, proposalHash []byte, checkHash bool) (bool, []byte, exitcode.ExitCode) {
+func (a Actor) approveTransaction(rt vmr.Runtime, txnID TxnID, txn Transaction) (bool, []byte, exitcode.ExitCode) {
 	var st State
-	var txn Transaction
-	rt.State().Transaction(&st, func() interface{} {
-		var err error
-		txn, err = st.getPendingTransaction(adt.AsStore(rt), txnID)
-		if err != nil {
-			rt.Abortf(exitcode.ErrNotFound, "failed to get transaction for approval: %v", err)
-		}
-		// abort duplicate approval
-		for _, previousApprover := range txn.Approved {
-			if previousApprover == rt.Message().Caller() {
-				rt.Abortf(exitcode.ErrIllegalState, "already approved this message")
-			}
-		}
-
-		// confirm the hashes match
-		if checkHash {
-			calculatedHash, err := ComputeProposalHash(&txn, rt.Syscalls().HashBlake2b)
-			if err != nil {
-				rt.Abortf(exitcode.ErrIllegalState, "failed to compute proposal hash: %v", err)
-			}
-			if proposalHash != nil && !bytes.Equal(proposalHash, calculatedHash[:]) {
-				rt.Abortf(exitcode.ErrIllegalState, "hash does not match proposal params")
-			}
+	// abort duplicate approval
+	for _, previousApprover := range txn.Approved {
+		if previousApprover == rt.Message().Caller() {
+			rt.Abortf(exitcode.ErrIllegalState, "already approved this message")
 		}
+	}
 
+	// add the caller to the list of approvers
+	rt.State().Transaction(&st, func() interface{} {
 		// update approved on the transaction
 		txn.Approved = append(txn.Approved, rt.Message().Caller())
-		if err = st.putPendingTransaction(adt.AsStore(rt), txnID, txn); err != nil {
+		if err := st.putPendingTransaction(adt.AsStore(rt), txnID, txn); err != nil {
 			rt.Abortf(exitcode.ErrIllegalState, "failed to put transaction for approval: %v", err)
 		}
 		return nil
 	})
 
+	return executeTransactionIfApproved(rt,st, txnID, txn)
+}
+
+func (a Actor) getTransaction(rt vmr.Runtime, st State, txnID TxnID, proposalHash []byte, checkHash bool) Transaction {
```

Comment:
There's a bit much copying of `Transaction` structs now. Return a pointer (never nil) from here, and other places where a Transaction is returned or taken as parameter. Change `st.putPendingTransaction` to accept a pointer too.

Reply:
This has been done.